standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+	- the current total rate limit MUST be decremented by the rate limit of membership B;
+	- the contract MUST take all necessary steps to ensure that the holder of membership B can withdraw their deposit later;
+- registration MUST fail if the total rate limit of _Active_, _GracePeriod_, and _Expired_ memberships, including the one being created, would exceed the maximum total rate;
+- registration MUST fail if the requested rate limit for a new membership is lower than the minimal allowed rate limit;
+- the user MUST lock-up a deposit to register a membership;
+- the user MUST specify the rate limit of the new membership[^2];
+- the size of the deposit MUST depend on the requested rate limit;
+- in case of a successful registration:
+	- the new membership MUST become _Active_;
+	- the current total rate limit MUST be incremented by the rate limit of the new membership;
+- a newly created membership MUST have an expiration time `T` and a grace period `G`.
+
+[^2]: A user-facing application SHOULD suggest default values for rate limits for the user.
+
+### Extend a membership
+
+Extending a membership is subject to the following conditions:
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership holder MUST be able to extend their membership;
+- any user except the membership holder MUST NOT be able to extend a membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Holding a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
+
+### Withdraw the deposit
+
+Deposit withdrawal is subject to the following conditions:
+- the membership holder MUST be able to withdraw their deposit;
+- any user except the membership holder MUST NOT be able to withdraw its deposit;
+- a deposit MUST be withdrawn in full;
+- a withdrawal MUST fail if the membership is not in _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_;
+- a membership MUST become _Erased_ after withdrawal.
+
+## Governance and upgradability
+
+At initial mainnet deployment, the contract MUST have an _Owner_.
+The _Owner_ MUST be able to change the values of all contract parameters.
+The _Owner_ MUST be able to pause any of the following contract functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+At some point, the _Owner_ SHOULD renounce their privileges, and the contract MUST become immutable.
+Further upgrades, if necessary, SHOULD be done by deploying a new contract and migrating the membership set.
+
+## Implementation Suggestions
+
+User-facing application SHOULD suggest a few rate limits (tiers) to simplify their users' choice.
+The RECOMMENDED rate limits in a three-tier model are as follows:
+- `20` messages per epoch as low-tier;
+- `200` messages per epoch as mid-tier;
+- `600` messages per epoch as high-tier.
+Users SHOULD be able to select a custom rate limit under advanced settings.
```

Comment:
```suggestion
User-facing application SHOULD suggest one or a few rate limits (tiers) to simplify their users' choice among the following RECOMMENDED rate limits:
- `20` messages per epoch as low-tier;
- `200` messages per epoch as mid-tier;
- `600` messages per epoch as high-tier.
```

I wouldn't expect an application to display all 3 tiers to all users. Really depends on the context.

Also, the recommendation of rate limit is done. It does imply that an application developer can propose rate limits outside of this 3 and just need to consider it carefully.

Reply:
I'm not sure I understand your suggestion. Do you suggest that an app SHOULD suggest one or multiple tiers to users, which may or may not intersect with our three RECOMMENDED tiers? Do you also think that a user SHOULD also, as an alternative to suggested tiers, be able to pick an arbitrary rate limit, perhaps under "advanced settings"?

Reply:
My suggestion is as per the `suggested change` above.

1. I don't think we should recommend app developers to show all 3 tiers to users. It really depends on the use case, and more choice you offer to a user, more friction you create. On the contrary, I would recommend app developer to choose one rate to display to the user, depending on the application. The application developer should think abotu it and select a rate among the recommended one.
2. I don't think user should be able to pick an arbitrary rate limit, because again, more choice more friction.